rfc-134-govuk-wide-session-cookie-and-login.md

```diff
@@ -0,0 +1,337 @@
+appropriate session duration with product and security input.
+
+### Create a new app to manage the auth process
+
+It's weird to have the login and logout controllers for GOV.UK as a
+whole located under `/transition-check`.
+
+We will create a new app—called account-frontend, which will live on a
+new machine class called personalisation—to serve those controllers
+and to handle OAuth interactions with the GOV.UK Account
+system. Signing in will set the `GOVUK-Session-ID` response header,
+signing out will set the `GOVUK-End-Session` response header.
+
+This app will serve these public endpoints:
+
+- `GET /sign-in`: initiates the OAuth flow with the accounts system
+  and sends the user on a consent journey.  Accepts these parameters:
+
+    - `_ga`: cross-domain tracking parameter to pass to the accounts
+      domain (optional)
+    - `redirect_path`: path on GOV.UK to redirect back to after
+      authenticating (optional, default: `/`)
+    - `state`: see below (optional)
+
+- `GET /sign-in/callback`: where the accounts system sends the user
+  back to.  Sets the `GOVUK-Session-ID` response header and stores the
+  returned OAuth access and refresh tokens in the app's database, if
+  the user successfully signed in, and redirects the user back to the
+  redirect_path.
+
+- `GET /sign-out`: sets the `GOVUK-End-Session` response header,
+  deletes the OAuth tokens from the app's database, and redirects the
+  user to `/`.
+
+The app will also serve these internal endpoints:
+
+- `GET /api/attributes`: looks up and returns some attributes from the
+  user's account.  Accepts these parameters:
+
+    - `session_id`: the value of the `GOVUK-Session-ID` header
+    - `attributes`: list of attribute names
+
+    Returns a hash of attribute names and values.
+
+- `POST /api/attributes`: sets some attribute values on the user's
+  account.  Accepts these parameters:
+
+    - `session_id`: the value of the `GOVUK-Session-ID` header
+    - `attributes`: hash of attribute names and values
+
+  This is a partial update.  Attributes *not* named in the hash keep
+  their previous values.
+
+- `POST /api/state`: sets some attribute values that will be persisted
+  if the user creates a new account, regardless of whether the user
+  returns to GOV.UK.  Accepts these parameters:
+
+    - `attributes`: hash of attribute names and values
+
+    Returns an ID which can be passed to `/sign-in`.
+
+Here are a few examples of how the Transition Checker will work:
+
+#### Clicking the "Sign in" link in the header
+
+1. The link sends the user to `/sign-in?redirect_path=...&_ga=...`
+
+2. The user is redirected to `www.account.publishing....`
+
+3. The user authenticates (registers or signs in)
+
+4. The user is redirected to `www.gov.uk/sign-in/callback?...`
+
+5. A session ID is generated, and used to persist the OAuth access and
+   refresh tokens to the app's database
+
+6. The `GOVUK-Session-ID` response header is set
```

Comment:
A few questions come to mind about the intentions of an app once they have one of these?

I assume an app can't assume that the presence of this header means that a user is signed in? is that something that needs to be confirmed with the account frontend? Would this be confirmed on each request or just once in a period of time? 

Would apps remain having their own sessions (eg _finder_frontend_session) (this seems essential for things like flash messages) and would there be any intersection with this govuk_session and those?

Reply:
If there's a `GOVUK-Session-ID`, then the user is signed in, because they have a valid session cookie.

However, it's not a *user* ID: all you can do with the session ID is get or set attributes, by querying the endpoints in the (new?) app.  It could be that the OAuth token associated with the session ID has been revoked, in which case any attempt to get/set attributes would fail (a 401 status seems most appropriate).  However if an app wants to display different information to signed in vs signed out users, then it can just use the presence of the session ID without checking attributes.

In fact that's how the account navigation in the header on [/transition](https://www.gov.uk/transition) works: it just checks for the presence of a finder-frontend session cookie.  It doesn't try to read the session cookie because /transition is rendered by collections.

Apps would keep their own sessions for flash messages and the like, if needed.

Reply:
Right - gotcha. Thanks for explaining. So because of the verification there are assumptions we can make about the session id's validity (because of the signature check) and to do anything substantial there needs to be an API call that's actually authenticated. I guess by that logic it's preferential to not cache any attributes/account stuff in the local session as that might have a very different lifetime.

Reply:
With the latest change an app can't assume that because there is a header it's valid; but it can use the public key to check that.